[UNI-295] Integration test for model_runner_2 and param_finder

ywcui1990 opened the pull request:
@marionleborgne @vitaly-krugl 

Integration test for param_finder_runner and model_runner_2

Summary
* Make sure we get expected model parameters from param_finder on 2 NAB datasets
* Make sure we get valid anomaly scores from model_runner_2

vitaly-krugl commented:
@ywcui1990 @marionleborgne, the new compatibility test fails when running from your branch on my macbookpro:

The first time I ran it, `testNYCTaxi` test failed. Then, I uninstalled and reinstalled unicorn_backend prior to running the test the second time; this resulted in download of a new version of nupic 0.3.6; both tests failed after this:

`$py.test model_runner_2_compatibility_test.py`

```
self = <model_runner_2_compatibility_test.ModelRunnerCompatibilityTest testMethod=testAmbientTemperatureSystemFailure>
name = 'ambient_temperature_system_failure'
inputSpec = '{"datetimeFormat": "%Y-%m-%d %H:%M:%S", "timestampIndex": 0, "csv": "/Users/vkruglikov/nta/numenta-apps/unicorn/tests/py/integration/compatibility_test/data/ambient_temperature_system_failure.csv", "rowOffset": 4, "valueIndex": 1}'
aggSpec = '{"windowSize": 26150, "func": "mean"}'
modelSpec = '{"modelId": "test", "inferenceArgs": {"predictionSteps": [1], "predictedField": "c1", "inputPredictedField": "auto"},...le": true, "inferenceType": "TemporalAnomaly", "tpEnable": true}}, "valueFieldName": "c1", "timestampFieldName": "c0"}'

    def _testModelRunner(self, name, inputSpec, aggSpec, modelSpec):
      with self._startModelRunnerSubprocess(
        inputSpec, aggSpec, modelSpec) as mrProcess:
    
        stdoutData, stderrData = mrProcess.communicate()
        out = stdoutData.splitlines()
        self.assertEqual(stderrData, "")
    
        results = self._load(os.path.join(RESULTS_DIR, name+'.csv'))
    
        for i in xrange(len(out)):
          outputRecord = json.loads(out[i])
    
          trueDataValue = float(results[i][1])
          trueAnomalyLikelihood = float(results[i][2])
          self.assertAlmostEqual(outputRecord[1], trueDataValue)
>         self.assertAlmostEqual(outputRecord[2], trueAnomalyLikelihood)
E         AssertionError: 0.0036569507561679745 != 0.013921482177631672 within 7 places
```

and

```
self = <model_runner_2_compatibility_test.ModelRunnerCompatibilityTest testMethod=testNYCTaxi>, name = 'nyc_taxi'
inputSpec = '{"datetimeFormat": "%Y-%m-%d %H:%M:%S", "timestampIndex": 0, "csv": "/Users/vkruglikov/nta/numenta-apps/unicorn/tests/py/integration/compatibility_test/data/nyc_taxi.csv", "rowOffset": 4, "valueIndex": 1}'
aggSpec = '{"windowSize": 7200, "func": "mean"}'
modelSpec = '{"modelId": "test", "inferenceArgs": {"predictionSteps": [1], "predictedField": "c1", "inputPredictedField": "auto"},...le": true, "inferenceType": "TemporalAnomaly", "tpEnable": true}}, "valueFieldName": "c1", "timestampFieldName": "c0"}'

    def _testModelRunner(self, name, inputSpec, aggSpec, modelSpec):
      with self._startModelRunnerSubprocess(
        inputSpec, aggSpec, modelSpec) as mrProcess:
    
        stdoutData, stderrData = mrProcess.communicate()
        out = stdoutData.splitlines()
        self.assertEqual(stderrData, "")
    
        results = self._load(os.path.join(RESULTS_DIR, name+'.csv'))
    
        for i in xrange(len(out)):
          outputRecord = json.loads(out[i])
    
          trueDataValue = float(results[i][1])
          trueAnomalyLikelihood = float(results[i][2])
          self.assertAlmostEqual(outputRecord[1], trueDataValue)
>         self.assertAlmostEqual(outputRecord[2], trueAnomalyLikelihood)
E         AssertionError: 0.01602313929629138 != 0.05106919905738879 within 7 places

model_runner_2_compatibility_test.py:152: AssertionError
```

vitaly-krugl commented on unicorn/tests/py/integration/compatibility_test/model_runner_2_compatibility_test.py:
```diff
@@ -0,0 +1,184 @@
+
+
+  @staticmethod
+  def _load(path):
```
Load what? This code probably inherited the name from the older version of the test, but it's still a good idea to improve. Perhaps `_loadCsvFile(path)`

vitaly-krugl commented on unicorn/tests/py/integration/compatibility_test/model_runner_2_compatibility_test.py:
```diff
@@ -0,0 +1,184 @@
+
+  def _testParamFinderRunner(self, name, inputSpec):
+
+    with self._startParamFinderRunnerSubprocess(inputSpec) as mrProcess:
```
The "mr" prefix in `mrProcess` referes to model runner. Since this function is starting the param finder (not model runner), this suffix is inappropriate.

vitaly-krugl commented on unicorn/tests/py/integration/compatibility_test/model_runner_2_compatibility_test.py:
```diff
@@ -0,0 +1,184 @@
+      self.assertEqual(stderrData, "")
+
+      results = self._load(os.path.join(RESULTS_DIR, name+'.csv'))
+
```
There should also be an assert that checks the sizes of the lists:

```
self.assertEqual(len(out), len(results))
```

Assuming that they are supposed to have the same number or rows.

vitaly-krugl commented on unicorn/tests/py/integration/compatibility_test/model_runner_2_compatibility_test.py:
```diff
@@ -0,0 +1,184 @@
+
+      results = self._load(os.path.join(RESULTS_DIR, name+'.csv'))
+
+      for i in xrange(len(out)):
```
A more pythonic way to structure this and avoid the unnecessary indexing is like this:

```
for computedResult, actualResult in zip(out, result):
  computedResult = jsonloads(computedResult)

  . . .

vitaly-krugl commented on unicorn/tests/py/integration/compatibility_test/model_runner_2_compatibility_test.py:
```diff
@@ -0,0 +1,184 @@
+      self.assertEqual(mrProcess.returncode, 0)
+
+
+  def _testParamFinderAndModelRunner(self, name):
```
Need docstrings here and elsewhere. What is "name"? Need `:param ... name: ...`

vitaly-krugl commented:
@ywcui1990, I am done with the code review. Both tests fail as described in my earlier comment, and there is additional feedback.

ywcui1990 commented:
@vitaly-krugl @marionleborgne 
Thanks for the code review @vitaly-krugl I was able to reproduce the test failure after I reinstalled unicorn. I am working on the issue now.

ywcui1990 commented:
@vitaly-krugl @marionleborgne 
I have fixed the compatibility test and included Vitaly's feedback. 

It seems that the test failure was due to some recent changes in nupic. I have updated the "correct" results using the latest nupic. I visually examined the current result with the anomaly likelihood I had before. They have similar shape, but not identical.

vitaly-krugl commented:
@ywcui1990 - some minor finds this time; please seem my new comments. LGTM and merge once resolved or deemed not a problem.

ywcui1990 commented:
@vitaly-krugl Thanks. I have merged it. https://github.com/numenta/nupic/pull/3047 shouldn't break the test as far as I can see.
